# Lab | Pandas

In this lab, we will be working with the customer data from an insurance company, which can be found in the CSV file located at the following link: https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

The data includes information such as customer ID, state, gender, education, income, and other variables that can be used to perform various analyses.

Throughout the lab, we will be using the pandas library in Python to manipulate and analyze the data. Pandas is a powerful library that provides various data manipulation and analysis tools, including the ability to load and manipulate data from a variety of sources, including CSV files.

### Data Description

- Customer - Customer ID

- ST - State where customers live

- Gender - Gender of the customer

- Education - Background education of customers 

- Customer Lifetime Value - Customer lifetime value(CLV) is the total revenue the client will derive from their entire relationship with a customer. In other words, is the predicted or calculated value of a customer over their entire duration as a policyholder with the insurance company. It is an estimation of the net profit that the insurance company expects to generate from a customer throughout their relationship with the company. Customer Lifetime Value takes into account factors such as the duration of the customer's policy, premium payments, claim history, renewal likelihood, and potential additional services or products the customer may purchase. It helps insurers assess the long-term profitability and value associated with retaining a particular customer.

- Income - Customers income

- Monthly Premium Auto - Amount of money the customer pays on a monthly basis as a premium for their auto insurance coverage. It represents the recurring cost that the insured person must pay to maintain their insurance policy and receive coverage for potential damages, accidents, or other covered events related to their vehicle.

- Number of Open Complaints - Number of complaints the customer opened

- Policy Type - There are three type of policies in car insurance (Corporate Auto, Personal Auto, and Special Auto)

- Vehicle Class - Type of vehicle classes that customers have Two-Door Car, Four-Door Car SUV, Luxury SUV, Sports Car, and Luxury Car

- Total Claim Amount - the sum of all claims made by the customer. It represents the total monetary value of all approved claims for incidents such as accidents, theft, vandalism, or other covered events.


External Resources: https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9

## Challenge 1: Understanding the data

In this challenge, you will use pandas to explore a given dataset. Your task is to gain a deep understanding of the data by analyzing its characteristics, dimensions, and statistical properties.

- Identify the dimensions of the dataset by determining the number of rows and columns it contains.
- Determine the data types of each column and evaluate whether they are appropriate for the nature of the variable. You should also provide suggestions for fixing any incorrect data types.
- Identify the number of unique values for each column and determine which columns appear to be categorical. You should also describe the unique values of each categorical column and the range of values for numerical columns, and give your insights.
- Compute summary statistics such as mean, median, mode, standard deviation, and quartiles to understand the central tendency and distribution of the data for numerical columns. You should also provide your conclusions based on these summary statistics.
- Compute summary statistics for categorical columns and providing your conclusions based on these statistics.

In [ ]:
# Your code here

# set-up, load the dataset

# dataset location:  https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv

import pandas as pd

url = "https://raw.githubusercontent.com/data-bootcamp-v4/data/main/file1.csv"
data = pd.read_csv(url)
print(data.head())


  Customer          ST GENDER             Education Customer Lifetime Value  \
0  RB50392  Washington    NaN                Master                     NaN   
1  QZ44356     Arizona      F              Bachelor              697953.59%   
2  AI49188      Nevada      F              Bachelor             1288743.17%   
3  WW63253  California      M              Bachelor              764586.18%   
4  GA49547  Washington      M  High School or Below              536307.65%   

    Income  Monthly Premium Auto Number of Open Complaints     Policy Type  \
0      0.0                1000.0                    1/0/00   Personal Auto   
1      0.0                  94.0                    1/0/00   Personal Auto   
2  48767.0                 108.0                    1/0/00   Personal Auto   
3      0.0                 106.0                    1/0/00  Corporate Auto   
4  36357.0                  68.0                    1/0/00   Personal Auto   

   Vehicle Class  Total Claim Amount  
0  Four-Door Car 

In [3]:
# Identify the dimensions of the dataset 
# by determining the number of rows and columns it contains.
print("Dataset dimensions:")
print(data.shape)

Dataset dimensions:
(4008, 11)


> **Conclusion Dataset Dimensions:**
>
> The dataset contains 4,008 rows and 11 columns, which represents a medium-sized dataset.
>
> From a purely quantitative perspective, this size is appropriate for exploratory data analysis using pandas and does not present performance limitations.
>
> In principle, the number of rows appears sufficient to extract meaningful statistical insights and identify patterns. However, the reliability of these insights strongly depends on data quality. Issues such as inconsistent categorical values, missing data, incorrect data types, or outliers may reduce the validity of the results. Therefore, data cleaning and validation are necessary steps before drawing definitive conclusions or using the data for further analysis or modeling.

In [4]:
# Determine the data types of each column 
# and evaluate whether they are appropriate for the nature of the variable. 

print("\nData types of each column:")
print(data.dtypes)



Data types of each column:
Customer                      object
ST                            object
GENDER                        object
Education                     object
Customer Lifetime Value       object
Income                       float64
Monthly Premium Auto         float64
Number of Open Complaints     object
Policy Type                   object
Vehicle Class                 object
Total Claim Amount           float64
dtype: object


In [6]:
# Identify the number of unique values for each column 
# and determine which columns appear to be categorical. 
# You should also describe the unique values of each categorical column 
# and the range of values for numerical columns, and give your insights.

print("\nUnique values and ranges for each column:")
for column in data.columns:
    unique_values = data[column].nunique()
    print(f"\nColumn: {column}")
    print(f"Number of unique values: {unique_values}")
    if data[column].dtype == 'object':
        print("Categorical values:")
        print(data[column].value_counts())
    else:
        print("Numerical range:")
        print(f"Min: {data[column].min()}, Max: {data[column].max()}")


Unique values and ranges for each column:

Column: Customer
Number of unique values: 1071
Categorical values:
Customer
MY31220    1
RB50392    1
QZ44356    1
AI49188    1
WW63253    1
          ..
FL50705    1
WC83389    1
OE15005    1
FV94802    1
BW63560    1
Name: count, Length: 1071, dtype: int64

Column: ST
Number of unique values: 8
Categorical values:
ST
Oregon        320
California    211
Arizona       186
Cali          120
Nevada         98
Washington     81
WA             30
AZ             25
Name: count, dtype: int64

Column: GENDER
Number of unique values: 5
Categorical values:
GENDER
F         457
M         413
Male       39
female     28
Femal      17
Name: count, dtype: int64

Column: Education
Number of unique values: 6
Categorical values:
Education
Bachelor                324
College                 313
High School or Below    296
Master                   94
Doctor                   37
Bachelors                 7
Name: count, dtype: int64

Column: Customer Lifetime Va

In [7]:
#  Compute summary statistics such as mean, median, mode, standard deviation, 
# and quartiles to understand the central tendency 
# and distribution of the data for numerical columns. 
# You should also provide your conclusions based on these summary statistics.
print("\nSummary statistics for numerical columns:")
print(data.describe())


Summary statistics for numerical columns:
             Income  Monthly Premium Auto  Total Claim Amount
count   1071.000000           1071.000000         1071.000000
mean   39295.701214            193.234360          404.986909
std    30469.427060           1601.190369          293.027260
min        0.000000             61.000000            0.382107
25%    14072.000000             68.000000          202.157702
50%    36234.000000             83.000000          354.729129
75%    64631.000000            109.500000          532.800000
max    99960.000000          35354.000000         2893.239678


In [ ]:
# Compute summary statistics for categorical columns 
# and providing your conclusions based on these statistics.

print("\nSummary statistics for categorical columns:")
for column in data.select_dtypes(include=['object']).columns:
    print(f"\nColumn: {column}")
    print(data[column].describe())



Summary statistics for categorical columns:

Column: Customer
count        1071
unique       1071
top       MY31220
freq            1
Name: Customer, dtype: object

Column: ST
count       1071
unique         8
top       Oregon
freq         320
Name: ST, dtype: object

Column: GENDER
count     954
unique      5
top         F
freq      457
Name: GENDER, dtype: object

Column: Education
count         1071
unique           6
top       Bachelor
freq           324
Name: Education, dtype: object

Column: Customer Lifetime Value
count           1068
unique          1027
top       251459.20%
freq               4
Name: Customer Lifetime Value, dtype: object

Column: Number of Open Complaints
count       1071
unique         6
top       1/0/00
freq         830
Name: Number of Open Complaints, dtype: object

Column: Policy Type
count              1071
unique                3
top       Personal Auto
freq                780
Name: Policy Type, dtype: object

Column: Vehicle Class
count              1

In [5]:
# provide suggestions for fixing any incorrect data types.
# For example, if a column representing dates is stored as a string,
# suggest converting it to a datetime format.
# Suggestions for fixing incorrect data types
for column in data.columns:
    if data[column].dtype == 'object':
        try:
            pd.to_datetime(data[column])
            print(f"Suggestion: Convert column '{column}' to datetime format.")
        except (ValueError, TypeError):
            pass
    elif data[column].dtype == 'int64' or data[column].dtype == 'float64':
        if data[column].nunique() < 10:
            print(f"Suggestion: Consider converting column '{column}' to categorical type.")

C:\Users\pauli\AppData\Local\Temp\ipykernel_14548\1388568433.py:8: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(data[column])
C:\Users\pauli\AppData\Local\Temp\ipykernel_14548\1388568433.py:8: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(data[column])
C:\Users\pauli\AppData\Local\Temp\ipykernel_14548\1388568433.py:8: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  pd.to_datetime(data[column])
C:\Users\pauli\AppData\Local\Temp\ipykernel_14548\1388568433.py:8: UserWarning: Could not infer format, so each element will be parsed individually, falli

# **Estás aquí. Antes de seguir revisa tus códigos anteriores. Queda pendiente primero visualizar los uniques values de las columnas que tienen muchos unique values y que no se pueden ver todos, visualizar una por una en celda separada para mayor reconocimiento de la data. Después revisar datatype y describir cómo es y dar recomendaciones de transformación. Después revisa todos los códigos, algunos ponen que has de dar tu conclusion sobre las estadisticas del estado de la tada, etce. Revisa si los códigos cumple lo que pide el enunciado y avanza al   
     challenge 2.

## Challenge 2: analyzing the data

### Exercise 1

The marketing team wants to know the top 5 less common customer locations. Create a pandas Series object that contains the customer locations and their frequencies, and then retrieve the top 5 less common locations in ascending order.

In [ ]:
# Your code here

### Exercise 2

The sales team wants to know the total number of policies sold for each type of policy. Create a pandas Series object that contains the policy types and their total number of policies sold, and then retrieve the policy type with the highest number of policies sold.

*Hint:*
- *Using value_counts() method simplifies this analysis.*
- *Futhermore, there is a method that returns the index of the maximum value in a column or row.*


In [ ]:
# Your code here

### Exercise 3

The sales team wants to know if customers with Personal Auto have a lower income than those with Corporate Auto. How does the average income compare between the two policy types?

- Use *loc* to create two dataframes: one containing only Personal Auto policies and one containing only Corporate Auto policies.
- Calculate the average income for each policy.
- Print the results.

In [ ]:
# Your code here

### Bonus: Exercise 4


Your goal is to identify customers with a high policy claim amount.

Instructions:

- Review again the statistics for total claim amount to gain an understanding of the data.
- To identify potential areas for improving customer retention and profitability, we want to focus on customers with a high policy claim amount. Consider customers with a high policy claim amount to be those in the top 25% of the total claim amount. Create a pandas DataFrame object that contains information about customers with a policy claim amount greater than the 75th percentile.
- Use DataFrame methods to calculate summary statistics about the high policy claim amount data. 

*Note: When analyzing data, we often want to focus on certain groups of values to gain insights. Percentiles are a useful tool to help us define these groups. A percentile is a measure that tells us what percentage of values in a dataset are below a certain value. For example, the 75th percentile represents the value below which 75% of the data falls. Similarly, the 25th percentile represents the value below which 25% of the data falls. When we talk about the top 25%, we are referring to the values that fall above the 75th percentile, which represent the top quarter of the data. On the other hand, when we talk about the bottom 25%, we are referring to the values that fall below the 25th percentile, which represent the bottom quarter of the data. By focusing on these groups, we can identify patterns and trends that may be useful for making decisions and taking action.*

*Hint: look for a method that gives you the percentile or quantile 0.75 and 0.25 for a Pandas Series.*

*Hint 2: check `Boolean selection according to the values of a single column` in https://towardsdatascience.com/filtering-data-frames-in-pandas-b570b1f834b9*

In [ ]:
# Your code here